agent/pipeline_uploader.go

```diff
@@ -0,0 +1,261 @@
+package agent
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/buildkite/agent/v3/api"
+	"github.com/buildkite/agent/v3/logger"
+	"github.com/buildkite/roko"
+)
+
+const (
+	defaultAttempts                 = 60
+	defaultSleepDuration            = 5 * time.Second
+	defaultSleepAfterUploadDuration = time.Second
+)
+
+var locationRegex = regexp.MustCompile(`jobs/(?P<jobID>[^/]+)/pipelines/(?P<uploadUUID>[^/]+)`)
+
+type PipelineUploader struct {
+	Client         APIClient
+	Change         *api.PipelineChange
+	JobID          string
+	RetrySleepFunc func(time.Duration)
+}
+
+func (u *PipelineUploader) AsyncUploadFlow(ctx context.Context, l logger.Logger) error {
```

Comment:
From the perspective of the other package, this is the only way to upload a pipeline with the new `PipelineUploader` (is that correct?) So I wonder if it should be named `Upload`. The inner workings make use of an async approach, but the whole method isn't exactly "async".

Reply:
Yep, the name was a holdover from my first attempt, where it did not fall back.